salt/pillar/nsot.py

```diff
@@ -0,0 +1,181 @@
+    query = salt.utils.http.query(url, data=data_dict, method='POST',
+                                  decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get auth_token from nsot! reason: %s', error)
+        return False
+    else:
+        log.debug('successfully obtained token from nsot!')
+        return query['dict']['auth_token']
+
+
+def _check_minion_regex(minion_id, minion_regex):
+    '''
+    check whether or not this minion should have this external pillar returned
+
+    :param minion_id: str
+    :param minion_regex: list
+    :return: bool
+    '''
+    get_pillar = False
+    for pattern in minion_regex:
+        log.debug('searching %s using %s', minion_id, minion_regex)
+        match = re.search(pattern, minion_id)
+        if match and match.string == minion_id:
+            log.debug('found match! %s: regex %s', minion_id, minion_regex)
+            get_pillar = True
+            break
+        log.debug('unable to match %s using regex %s', minion_id, minion_regex)
+    return get_pillar
+
+
+def _query_nsot(url, headers, device=None):
+    url += 'devices/'
+    if not device:
+        query = salt.utils.http.query(url, header_dict=headers, decode=True)
+    else:
+        device += '/'
+        query = salt.utils.http.query(url + device, header_dict=headers,
+                                      decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get device(s) from nsot! reason: %s', error)
+        return {}
+    else:
+        return query['dict']
+
+
+def ext_pillar(minion_id,
+               pillar,
+               api_url,
+               email,
+               secret_key,
+               replace_dot=None,
+               all_device_info=None,
+               minion_regex=None):
+    '''
+    Query NSoT API for network devices
+    '''
+    ret = {}
+    if minion_id == '*':
+        log.info('There\'s no data to collect from NSoT for the Master')
+        return ret
+
+    if minion_regex:
+        get_ext_pillar = _check_minion_regex(minion_id, minion_regex)
+        if not get_ext_pillar:
+            if all_device_info:
+                if minion_id not in all_device_info:
+                    return ret
+            else:
+                return ret
+
+    token = _get_token(api_url, email, secret_key)
+
+    if not token:
+        return ret
+
+    headers = {'Authorization': 'AuthToken {}:{}'.format(email, token)}
+
+    if all_device_info:
+        if minion_id in all_device_info:
+            all_devices = _query_nsot(api_url, headers)
+            if all_devices:
+                ret['nsot'] = {}
+                ret['nsot']['devices'] = all_devices
+            return ret
+
+    if replace_dot:
+        minion_id = minion_id.replace('.', replace_dot)
+
+    device_info = _query_nsot(api_url, headers, device=minion_id)
+    if device_info:
+        ret['nsot'] = device_info
+
+    return ret
```

Comment:
This is not very nice. Please make sure we have only one return at the end, so we can continue maintaining this code in a future.